File: solutions/Two Pointers/42_trapping-rain-water.py
from typing import List

class Solution:
    def trap(self, height: List[int]) -> int:
        l, r = 0, len(height) -1
        ll, rl = height[l], height[r]
        res = 0

        while l < r:
            if ll < rl:
                res += ll - height[l]
                l += 1
                ll = max(ll, height[l])

            else:
                res += rl - height[r]
                r -= 1
                rl = max(rl, height[r])

        return res
    
# Time Complexity: O(n)
# Space Complexity: O(1)


class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        
        left = [0] * (n+1)
        for i in range(n):
            left[i + 1] = max(height[i], left[i])

        right = [0] * (n+1)
        for i in range(n -2, -1, -1):
            right[i] = max(right[i + 1], height[i + 1])

        res = 0
        for i in range(n):
            res += max(0, min(left[i], right[i]) - height[i])

        return res
    
# Time Complexity: O(n)
# Space Complexity: O(n)

File: C++/Two Pointers/42_trapping-rain-water.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int trap(vector<int>& height) {
        size_t n = height.size();
        int res = 0;

        vector<int> prefix(n + 1, 0);
        for (int i = 0; i < n; ++i) {
            prefix[i + 1] = max(prefix[i], height[i]);
        }

        vector<int> suffix(n + 1, 0);
        for (int i = n - 1; i >= 0; --i) {
            suffix[i] = max(suffix[i + 1], height[i]);
        }

        for (int i = 0; i < n; ++i) {
            res += max(0, min(prefix[i], suffix[i + 1]) - height[i]);
        }
        
        return res;
    }
};


class Solution {
public:
    int trap(vector<int>& height) {
        size_t n = height.size();
        int res = 0;
        int l = 0, r = n - 1;
        int lmax, rmax;

        while (l < r) {
            lmax = max(lmax, height[l]);
            rmax = max(rmax, height[r]);

            if (height[l] < height[r]) {
                res += lmax - height[l];
                l++;

            } else {
                res += rmax - height[r];
                r--;
            }
        }

        return res;
    }
};